[Java] Add IAS Support to Service Binding Documentation (#1755)
SAP/cloud-sdk: docs-java/features/connectivity/003-service-bindings.mdx

  - btp
---

> [_Service Bindings_](https://github.com/SAP/btp-environment-variable-access/wiki/Fundamentals#service-binding) are a fundamental concept of application deployment on the SAP Business Technology Platform.

Author: This is still explained in the advanced section. But I felt here it is more natural to describe it from the perspective of the user of "I want to connect to a bound service" and less from the perspective of how we designed the API "convert ServiceBinding -> HttpDestination`"

## Connecting to a Service

You can obtain a `Destination` to connect to a service by using the `ServiceBindingDestinationLoader` API:

```java
var options = ServiceBindingDestinationOptions
        .forService(ServiceIdentifier.<MY_SERVICE>)
        .build();

HttpDestination destination = ServiceBindingDestinationLoader.defaultLoaderChain()
        .getDestination(options);
```

Choose from the available `ServiceIdentifier` constants to specify the service you want to connect to.

For example, for connecting to the [Workflow REST API](https://api.sap.com/api/SAP_CP_Workflow_CF/overview) use:

```java {2-3}
var options = ServiceBindingDestinationOptions
        .forService(ServiceIdentifier.WORKFLOW)
        .withOption(BtpServiceOptions.WorkflowOptions.REST_API)
        .build();

HttpDestination destination = ServiceBindingDestinationLoader.defaultLoaderChain()
        .getDestination(options);
```

Note the additional option `BtpServiceOptions.WorkflowOptions.REST_API` that is specific to the [Workflow API (REST)](https://api.sap.com/api/SAP_CP_Workflow_CF/overview).

> Explore the [BtpServiceOptions](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/BtpServiceOptions.html) class to find any options relevant for your service.

Reviewer: ```suggestion
For BTP services that offer different APIs (such as the Workflow service) you should also additionally define which API the resulting `HttpDestination` should point to using `BtpServiceOptions`.

Explore the [BtpServiceOptions](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/BtpServiceOptions.html) class to find the options relevant for your service and your use-case.
```

Author: I tweaked the suggestion a bit, because the options are not only for selecting an endpoint. So I rephrased it a bit to make it one example.

</details>

:::tip Your service is not in the list?

In case the service you want to connect to is not yet recognized by the SAP Cloud SDK by default, you can still use the API by providing some additional information.

For services backed by the SAP Identity and Authentication Service (IAS), please see the [section on IAS](#using-the-identity-and-authentication-service-ias) below.
For other services, please refer to the section on [adding support for more services](#adding-support-for-more-services).

:::

In case you are a service provider and want to make your service compatible with the SAP Cloud SDK: Please consider contributing!

- For IAS-based services it is sufficient to conform to the dedicated [service binding format](#service-binding-format) for IAS.
- For XSUAA-based services please refer to the section on [adding support for more services](#adding-support-for-more-services) and consider adding your implementation [here](https://github.com/SAP/cloud-sdk-java/blob/main/cloudplatform/connectivity-oauth/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/BtpServicePropertySuppliers.java) by raising a pull request.

### Working with Multiple Service Bindings

> In case your application is bound to multiple instances of the same service passing the `ServiceIdentifier` is not sufficient.

Reviewer: ```suggestion
In case your application is bound to multiple instances of the same service, passing the `ServiceIdentifier` is not sufficient.
```

    .build();
```

## Using the Identity and Authentication Service (IAS)

:::warning Beta API

The API for connecting to services secured by the SAP Identity and Authentication Service (IAS) is currently in beta and subject to change.

:::

In case your application is bound to an instance of the SAP Identity and Authentication Service (IAS) you can use the SAP Cloud SDK to connect to other applications and services that are secured using IAS.
Effectively, the SAP Cloud SDK implements the OAuth flows described [here](https://help.sap.com/docs/identity-authentication/identity-authentication/consume-apis-from-other-applications).

:::info Supported Credential Types

The SAP Cloud SDK supports the credential types `binding-secret` and `X509_GENERATED` for IAS service bindings.

<!-- Prepared text, to be enabled with https://github.com/SAP/cloud-sdk/pull/1756

> If you want to use the `X509_ATTESTED` credential type, you need to add the `connectivity-ztis` dependency to your project.

Reviewer: ```suggestion
If you want to use the `X509_GENERATED` credential type, you need to add the `connectivity-ztis` dependency to your project.
```

-->

The types `X509_PROVIDED` and `X509_ATTESTED` are currently not supported.
:::

### Connecting to Services

If your service is secured using IAS and is using the dedicated [service binding format](#service-binding-format) supported by the SAP Cloud SDK, you can obtain a destination by passing the service label as the `ServiceIdentifier`:

```java
var options = ServiceBindingDestinationOptions
        .forService(ServiceIdentifier.of("your-service-label"))
        .build();
```

In case your service is not using the default format you can still use the `IasOptions` to provide the necessary information:

```java
var options = ServiceBindingDestinationOptions
        .forService(ServiceIdentifier.IDENTITY_AUTHENTICATION)
        .withOption(IasOptions.withTargetUri("https://foo.com"))
        .build();
```

Obtain the URL from the service binding of the service you want to connect to.

:::info Mutual TLS Only Authentication

In case the service does not require a JWT token (e.g. the Event Broker service) you can pass the additional option `IasOptions.withoutTokenForTechnicalProviderUser()`.

> Together with setting `OnBehalfOf.TECHNICAL_USER_PROVIDER` this will skip retrieving a token from IAS and only sending the X.509 certificate, saving a bit of performance.

Reviewer: Giving a code snippet here instead would be better.

```java
var binding = new DefaultServiceBindingBuilder()
    .withServiceIdentifier(ZTIS_IDENTIFIER)
    .withCredentials(Map.of(
         "clientid", "id",
         "url", "https://foo.com"))
    .build();
```

You can pass the binding explicitly via the options as described in the section on [working with multiple bindings](#working-with-multiple-service-bindings).

However, in some cases the SAP Cloud SDK or other frameworks may implicitly access service bindings.
For such cases one can modify the global `ServiceBindingAccessor`.

The SAP Cloud SDK will use the default instance whenever looking for service bindings.
You can override this instance:

```java
DefaultServiceBindingAccessor.setInstance(() -> List.of(binding));
```

In case you want to keep all existing bindings and just add your own you can use the following code:

```java
final ServiceBindingAccessor accessor = DefaultServiceBindingAccessor.getInstance();
var newAccessor = new ServiceBindingMerger(List.of(accessor, () -> binding), ServiceBindingMerger.KEEP_EVERYTHING);
DefaultServiceBindingAccessor.setInstance(newAccessor);
```

> ## (Advanced) Background Information

Author: This section goes very in depth on how the API is designed and how it can be extended. For most users most of the details are not required, but it is very helpful whenever the default, simple cases are not enough (or for developers of the SDK itself ;) ).

I decided to take some of the basic information / simple cases and document them above. I cover the simplest possible case from the `### Customization` section into `## Adding Support for More Services`, but linking to the more advanced `### Customization` section for more information, since that contains a full (and thus complex) example.
